pkg/proxy/healthcheck/proxier_health.go

```diff
@@ -68,93 +66,114 @@ type proxierHealthServer struct {
 		listener:      listener,
 		httpFactory:   httpServerFactory,
 		clock:         c,
 		addr:          addr,
 		healthTimeout: healthTimeout,
-		recorder:      recorder,
-		nodeRef:       nodeRef,
+
+		lastUpdatedMap:         make(map[v1.IPFamily]time.Time),
+		oldestPendingQueuedMap: make(map[v1.IPFamily]time.Time),
+		// The node is eligible (and thus the proxy healthy) while it's starting up
+		// and until we've processed the first node event that indicates the
+		// contrary.
+		nodeEligible: true,
 	}
-	// The node is eligible (and thus the proxy healthy) while it's starting up
-	// and until we've processed the first node event that indicates the
-	// contrary.
-	hs.nodeEligible.Store(true)
-	return hs
 }
 
-// Updated indicates that kube-proxy has successfully updated its backend, so it should
-// be considered healthy now.
-func (hs *proxierHealthServer) Updated() {
-	hs.oldestPendingQueued.Store(zeroTime)
-	hs.lastUpdated.Store(hs.clock.Now())
+// Updated indicates that proxier for the given IP family has successfully updated
+// its backend, so it should be considered healthy now.
+func (hs *proxierHealthServer) Updated(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	hs.oldestPendingQueuedMap[ipFamily] = zeroTime
+	hs.lastUpdatedMap[ipFamily] = hs.clock.Now()
 }
 
-// QueuedUpdate indicates that the proxy has received changes from the apiserver but
-// has not yet pushed them to its backend. If the proxy does not call Updated within the
-// healthTimeout time then it will be considered unhealthy.
-func (hs *proxierHealthServer) QueuedUpdate() {
-	// Set oldestPendingQueued only if it's currently zero
-	hs.oldestPendingQueued.CompareAndSwap(zeroTime, hs.clock.Now())
+// QueuedUpdate indicates that the proxier for the given IP family has received changes
+// from the APIServer but has not yet pushed them to its backend. If the proxy does not
+// call Updated within the healthTimeout time then it will be considered unhealthy.
+func (hs *proxierHealthServer) QueuedUpdate(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	// Set oldestPendingQueuedMap only if it's currently zero
+	if val, ok := hs.oldestPendingQueuedMap[ipFamily]; ok && val == zeroTime {
+		hs.oldestPendingQueuedMap[ipFamily] = hs.clock.Now()
+	}
 }
 
 // IsHealthy returns only the proxier's health state, following the same
 // definition the HTTP server defines, but ignoring the state of the Node.
 func (hs *proxierHealthServer) IsHealthy() bool {
-	isHealthy, _, _ := hs.isHealthy()
+	isHealthy, _ := hs.isHealthy()
 	return isHealthy
 }
 
-func (hs *proxierHealthServer) isHealthy() (bool, time.Time, time.Time) {
-	var oldestPendingQueued, lastUpdated time.Time
-	if val := hs.oldestPendingQueued.Load(); val != nil {
-		oldestPendingQueued = val.(time.Time)
-	}
-	if val := hs.lastUpdated.Load(); val != nil {
-		lastUpdated = val.(time.Time)
-	}
+func (hs *proxierHealthServer) isHealthy() (bool, time.Time) {
+	hs.lock.RLock()
```

Comment:
@danwinship should I deep copy both the maps here and release the lock immediately ?

Reply:
no, this doesn't get called often enough to justify any heroic efforts at optimization.